samples/end-to-end-testing/README.md

```diff
@@ -0,0 +1,119 @@
+
+This is the common structure of a Golang test using Terratest:
+
+```Go
+package test
+
+import (
+    "testing"
+
+    "github.com/gruntwork-io/terratest/modules/terraform"
+    test_structure "github.com/gruntwork-io/terratest/modules/test-structure"
+)
+
+func TestEndToEndDeploymentScenario(t *testing.T) {
+    t.Parallel()
+
+    fixtureFolder := "../"
+
+    // User Terratest to deploy the infrastructure
+    test_structure.RunTestStage(t, "setup", func() {
+        terraformOptions := &terraform.Options{
+            // Indicate the directory that contains the Terraform configuration to deploy
+            TerraformDir: fixtureFolder,
+        }
+
+        // Save options for later test stages
+        test_structure.SaveTerraformOptions(t, fixtureFolder, terraformOptions)
+
+        // Triggers the terraform init and terraform apply command
+        terraform.InitAndApply(t, terraformOptions)
+    })
+
+    test_structure.RunTestStage(t, "validate", func() {
+        // run validation checks here
+        terraformOptions := test_structure.LoadTerraformOptions(t, fixtureFolder)
+		publicIpAddress := terraform.Output(t, terraformOptions, "public_ip_address")
+    })
+
+    // When the test is completed, teardown the infrastructure by calling terraform destroy
+    test_structure.RunTestStage(t, "teardown", func() {
+        terraformOptions := test_structure.LoadTerraformOptions(t, fixtureFolder)
+        terraform.Destroy(t, terraformOptions)
+    })
+}
+```
+
+As you can see in the snippet above, the test is composed by three stages:
+
+1. `setup`: this stage is responsible for running Terraform to deploy the configuration.
+2. `validate`: this stage is responsible for doing the validation checks / assertions.
+3. `teardown`: this stage is responsible for cleaning up the infrastructure.
+
+Some relevant functions provided by Terratest framework are:
+
+- `terraform.InitAndApply` allows to run the `terraform init` and `terraform apply` commands from Go code.
+- `terraform.Output` allows to retrieve the value of a deployment output variable.
+- `terraform.Destroy` allows to run the `terraform destroy` command from Go code.
+- `test_structure.LoadTerraformOptions` allows to load Terraform options (config, variables etc.) from the state.
+- `test_structure.SaveTerraformOptions` allows to save Terraform options (config, variables etc.) to the state.
+
+## Run the end-to-end test
+
+Running the test requires that Terraform is installed and configured on your machine and that you are connected to your Azure subscription with the Azure CLI command `az login`.
+
+Once ready, since the end-to-end test is just a Go test, it can be run like the following:
+
+```console
+# Set the path of the SSH private key to use to connect the virtual machine
+export TEST_SSH_KEY_PATH="/home/bob/.ssh/id_rsa"
+cd test
+go test -v ./ -timeout 10m
```

Comment:
Insufficien ttimeout settings make the test run fail very cryptic and dont raise that issue  - longer timeouts are advised depending on youresources